### DISCUSSION TITLE
String-based vs. Object-based file access

### DISCUSSION THREAD

**Discussion by atjn** [humanwhocodes/humanfs#11] [Feedback]
I think you made a mistake by basing access on text paths. I think a better way to handle this is via objects that represent a specific path. What if I want to access the `test.txt` file in the dir `files/`. I have to do string manipulation for that? I'd rather call `DirHandle.getFile("test.txt")` and get a `FileHandle` back that I can read/write to.

This is not just convenience, it is a common bug that string manipulation results in paths with double slashes, missing slashes, or one slash that is the wrong type (`\` vs `/`), or countless other normalization issues. Abstracting this away is IMO a service to the developer and the end user.

Object-based systems also have better type safety and much better intellisense.

Another reason to go with objects, is that the new file system APIs that are being implemented in browsers are object-based. In my experience, it is easy to adapt objects to path APIs, but hard to adapt paths to object APIs, especially given that the browser APIs require user permissions to be attached to the objects.

Maybe the object-based APIs can be added on top of fsx, but that seems like unnecessary complexity to me.

**Comment by atjn** [comment]
This is a repost of: https://mastodon.online/@atjn/111734301732772549 :)

**Comment by nzakas** [comment]
> Another reason to go with objects, is that the new file system APIs that are being implemented in browsers are object-based. In my experience, it is easy to adapt objects to path APIs, but hard to adapt paths to object APIs, especially given that the browser APIs require user permissions to be attached to the objects.

Can you give an example?

**Reply by atjn** [reply]
## Object-based API to String-based API

This is an example where fsx is modelled after the [origin private file system](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (object-based) and needs to access files via the [Node file system API](https://nodejs.org/api/fs.html) (string-based). Any operation in OPFS corresponds to a simple operation in NodeFS, so this is a simple translation.

|    | Object-based input (OPFS)                            | Resulting operation in fsx  (NodeFS)                                                                                                |
|----|------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------|
| 1. | Request DirectoryHandle D1 for access to a private directory | Create a new temporary directory, return a mock DirectoryHandle with a member variable _tempdir_ that is set to the path of the directory |
| 2. | Get DirectoryHandle D2 for "foo" in D1               | Return mock DirectoryHandle with new path                                                                                       |
| 3. | Get DirectoryHandle D3 for "bar" in D2               | Return mock DirectoryHandle with new path                                                                                       |
| 4. | Get FileHandle F1 for "file.txt" in D3               | Return mock FileHandle with new path                                                                                            |
| 5. | Write to F1                                          | Write to "{tempdir}/foo/bar/file.txt"                                                                                           |

## String-based API to Object-based API

This is an example where fsx is modelled after the [Node file system API](https://nodejs.org/api/fs.html) (string-based) and needs to access files via the  [origin private file system](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (object-based). Some operations in NodeFS do not have a corresponding operation in OPFS, so fsx needs to do a lot of work.

|    | String-based input (NodeFS)              | Resulting operation in fsx (OPFS)                                                                                                                                                                                                                              |
|----|-------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1. |                                           | Create a map M1 from strings (UUID) to DirectoryHandles                                                                                                                                                                                                 |
| 2. | Request the path P1 to a temporary folder | Request DirectoryHandle D1 for access to a private directory, generate a UUID as the name of the folder, save the UUID and the DirectoryHandle in M1 as a pair, return the UUID as the absolute path to the directory                                                     |
| 3. | Write to "{p1}/foo/bar/file.txt"              | Run a regex to find the name of the first folder, get the DirectoryHandle D1 that corresponds to that name from M1, get DirectoryHandle D2 for "foo" in D1, get DirectoryHandle D3 for "bar" in D2, get FileHandle F1 for "file.txt" in D3, write to F1 |
| 4. |                                           | You should probably cache D2, D3, and F1 to make subsequent operations faster, you will probably need to define different caching strategies for different scenarios                                                                                    |                                                                          |

You can never garbage collect any pair in M1 because you have no idea if the path to a directory is saved somewhere. You could implement some janky solution that allows the API to "delete" the folder. Then the next issue is what to do when encountering path names that are not in M1.

**Reply by nzakas** [reply]
Thanks, that's helpful. Overall I think OPFS is way too heavy-handed for most use cases and is really strictly designed to enable security between a browser and the local disk. There may be some learnings there, but I think my current approach is still the path I'll walk down.

**Reply by boneskull** [reply]
There are considerable benefits to the suggested idea _in the browser_, but when the underlying implementation (e.g., `node:fs`, `libuv` and the OS beneath it) doesn't support this paradigm, you won't get much bang for your implementation-buck.

_That being said_, it would be neat to have some sort of way to bring-your-own primitive.  Certainly, the user-facing implementations (e.g., `fsx-node`) don't need to expose it (or if they did, not by default).  It _would_ be neat if the core API accepted paths of type `T = string` instead of just `string`.  That way I could define a new instance with paths based on `T`; e.g., `T` is `URL` and now my instance only accepts `URL` paths.  

(Also: the core API should probably support `URL`, shouldn't it? I'd expect all target platforms to support that at this point)

**Reply by boneskull** [reply]
Also: `fsx` is interesting to me not because of the cross-platform-ness of it, but the swappable `Impl`s.  

If I own an `fsx` instance, _and_ I can provide an instance to my dependency, I can decide--at a high level, and with a considerable amount of granularity--what my dependency can do with it.  That makes it desirable from a supply-chain security standpoint.  I can choose to make e.g., `fsx.write()` a no-op for dependency `foo`, but leave it intact for `bar`.  There's far more to it than that (e.g., if a dependency tries to create its own `fsx` instance, it should only ever receive what I've allowed it to), but it's a start.

Of course, those dependencies must accept my `fsx` instance in the first place 😄.  Hence, #44.

**Reply by atjn** [reply]
> Overall I think OPFS is way too heavy-handed for most use cases and is really strictly designed to enable security between a browser and the local disk. There may be some learnings there, but I think my current approach is still the path I'll walk down.

That's fair, I am interested in seeing what you can do with it :) I don't see why OPFS is too heavy-handed, it seems to support all use cases but with better DX and with better type safety. You make it sound like the security aspect is too limiting for anything but browsers, but in my experience, native applications are looking at the security benefits of web apps and are trying to implement similar safety designs as well because it shields against so many mistakes. I firmly believe that object-based is the future of file system API's.

**Reply by nzakas** [reply]
At some point I'll write up why I don't think OPFS is a good option outside of the browser, but for now, it's clear that we disagree and that's okay. :smile:
